In [36]:
# Homework 9: SVM and MLP Model Comparison

## Ethan DeMelo 

In [37]:
# set up
import pandas as pd
import numpy as np
np.set_printoptions(formatter={'float_kind':"{:3.2f}".format})
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
 #Import accuracy score and balanced accuracy score from model metrics package
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score, balanced_accuracy_score
from sklearn.metrics import confusion_matrix,  roc_auc_score

df = pd.read_csv("https://raw.githubusercontent.com/Eldemelo/Classes/Development/CSC310/Homework/homework_8/CrohnD.csv")
df.head()

rownames   ID      nrAdvE   BMI     height   country   sex   age   weight   \
0          1   19908        4   25.22      163  c1        F       47       67   
1          2   19909        4   23.80      164  c1        F       53       64   
2          3   19910        1   23.05      164  c1        F       68       62   
3          4   20908        1   25.71      165  c1        F       48       70   
4          5   20909        2   25.95      170  c1        F       67       75   

     treat  
0  placebo  
1       d1  
2  placebo  
3       d2  
4  placebo

In [38]:
# Clean Data
df['country '] = df['country '].str.strip().replace('c1', '0')
df['country '] = df['country '].str.strip().replace('c2', '1')
df['sex '] = df['sex '].str.strip().replace('F', '0')
df['sex '] = df['sex '].str.strip().replace('M', '1')

df.head()

rownames   ID      nrAdvE   BMI     height  country  sex   age   weight   \
0          1   19908        4   25.22      163        0    0    47       67   
1          2   19909        4   23.80      164        0    0    53       64   
2          3   19910        1   23.05      164        0    0    68       62   
3          4   20908        1   25.71      165        0    0    48       70   
4          5   20909        2   25.95      170        0    0    67       75   

     treat  
0  placebo  
1       d1  
2  placebo  
3       d2  
4  placebo

Kernel: Linear

In [39]:
df = df.drop(['rownames '],axis=1)
X  = df.drop(['treat'],axis=1)
y = df['treat']

model = SVC(kernel='linear', C=0.001, max_iter=10000)

In [40]:
# do the 5-fold cross validation
scores = cross_val_score(model, X, y, cv=5)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.42 0.33 0.26 0.26 0.35]
Accuracy: 0.32


d:\Program Files (x86)\Python\lib\site-packages\sklearn\svm\_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
d:\Program Files (x86)\Python\lib\site-packages\sklearn\svm\_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
d:\Program Files (x86)\Python\lib\site-packages\sklearn\svm\_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
d:\Program Files (x86)\Python\lib\site-packages\sklearn\svm\_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=10000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
d:\Program Files (x86)\Python\lib\site-packages\sklearn\svm\_base.py:297: ConvergenceWarning: Solver

Kernel: Polynomial

In [41]:
df = pd.read_csv("https://raw.githubusercontent.com/Eldemelo/Classes/Development/CSC310/Homework/homework_8/CrohnD.csv")

# Clean Data
df['country '] = df['country '].str.strip().replace('c1', '0')
df['country '] = df['country '].str.strip().replace('c2', '1')
df['sex '] = df['sex '].str.strip().replace('F', '0')
df['sex '] = df['sex '].str.strip().replace('M', '1')

df = df.drop(['rownames '],axis=1)
X  = df.drop(['treat'],axis=1)
y = df['treat']

model = SVC(kernel='poly', degree=2, coef0=0.9)

# do the 5-fold cross validation
scores = cross_val_score(model, X, y, cv=5)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.33 0.33 0.30 0.30 0.30]
Accuracy: 0.32


Kernel: rbf

In [61]:
df = pd.read_csv("https://raw.githubusercontent.com/Eldemelo/Classes/Development/CSC310/Homework/homework_8/CrohnD.csv")

# Clean Data
df['country '] = df['country '].str.strip().replace('c1', '0')
df['country '] = df['country '].str.strip().replace('c2', '1')
df['sex '] = df['sex '].str.strip().replace('F', '0')
df['sex '] = df['sex '].str.strip().replace('M', '1')

df = df.drop(['rownames '],axis=1)
X  = df.drop(['treat'],axis=1)
y = df['treat']

model = SVC(kernel='rbf', gamma='scale')

# do the 5-fold cross validation
scores = cross_val_score(model, X, y, cv=5)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.33 0.33 0.30 0.30 0.30]
Accuracy: 0.32


Kernel: Sigmoid

In [60]:
df = pd.read_csv("https://raw.githubusercontent.com/Eldemelo/Classes/Development/CSC310/Homework/homework_8/CrohnD.csv")

# Clean Data
df['country '] = df['country '].str.strip().replace('c1', '0')
df['country '] = df['country '].str.strip().replace('c2', '1')
df['sex '] = df['sex '].str.strip().replace('F', '0')
df['sex '] = df['sex '].str.strip().replace('M', '1')

df = df.drop(['rownames '],axis=1)
X  = df.drop(['treat'],axis=1)
y = df['treat']

model = SVC(kernel='sigmoid', gamma='scale', coef0=0.8)

# do the 5-fold cross validation
scores = cross_val_score(model, X, y, cv=5)
print("Fold Accuracies: {}".format(scores))
print("Accuracy: {:3.2f}".format(scores.mean()))

Fold Accuracies: [0.33 0.54 0.30 0.30 0.30]
Accuracy: 0.36


By tuning the gamma parameter for both poly and sigmoid kernels, I've found that 'scale' yielded the best accuracy.

Looking at the 4 models, it appears that for this dataset the sigmoid model is the most accurate at 36%. The other 3 models yielded a 32% accuracy which is slightly worse. Overall, there is a significant difference between sigmoid and the rest, but there is no significant differnce between linear, poly, and rbf.